### ISSUE TITLE
IllegalStateException: Unknown status: CHANGE_STATUS_UNKNOWN 

### ISSUE THREAD

**Issue by npepinpe** [camunda/zeebe#15645]
**Describe the bug**

While the cluster was starting up, it seems like an unknown change status was propagated (or read?) to the other nodes. This caused the error (as expected). The cluster however recovered and was shortly running fine, so it seems the impact was quite low, as far as I can tell. _I did not test if further scaling operations still work though_.

It also only occurred on one cluster so far, so likelihood seems low.

**Expected behavior**

Either `CHANGE_STATUS_UNKNOWN` is expected and is not treated as an unexpected error case, or it is and it shouldn't have happened.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
 java.lang.IllegalStateException: Unknown status: CHANGE_STATUS_UNKNOWN
at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.toChangeStatus ( io/camunda.zeebe.topology.serializer/ProtoBufSerializer.java:667 )
at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decodeCompletedChange ( io/camunda.zeebe.topology.serializer/ProtoBufSerializer.java:342 )
at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decodeClusterTopology ( io/camunda.zeebe.topology.serializer/ProtoBufSerializer.java:109 )
at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decode ( io/camunda.zeebe.topology.serializer/ProtoBufSerializer.java:78 )
at io.atomix.cluster.messaging.impl.DefaultClusterCommunicationService$InternalMessageBiConsumer.accept ( io/atomix.cluster.messaging.impl/DefaultClusterCommunicationService.java:328 )
at io.atomix.cluster.messaging.impl.DefaultClusterCommunicationService$InternalMessageBiConsumer.accept ( io/atomix.cluster.messaging.impl/DefaultClusterCommunicationService.java:313 )
at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$registerHandler$6 ( io/atomix.cluster.messaging.impl/NettyMessagingService.java:277 )
at io.camunda.zeebe.scheduler.ActorJob.invoke ( io/camunda.zeebe.scheduler/ActorJob.java:94 )
at io.camunda.zeebe.scheduler.ActorJob.execute ( io/camunda.zeebe.scheduler/ActorJob.java:47 )
at io.camunda.zeebe.scheduler.ActorTask.execute ( io/camunda.zeebe.scheduler/ActorTask.java:120 )
at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask ( io/camunda.zeebe.scheduler/ActorThread.java:130 )
at io.camunda.zeebe.scheduler.ActorThread.doWork ( io/camunda.zeebe.scheduler/ActorThread.java:108 )
at io.camunda.zeebe.scheduler.ActorThread.run ( io/camunda.zeebe.scheduler/ActorThread.java:227 ) 
```

</p>
</details>

**Environment:**
- Zeebe Version: 8.4.0-alpha2

**Comment by oleschoenburg** [comment]
Happened again on int: https://console.cloud.google.com/errors/detail/CJPttavl2KffJw;service=zeebe;time=P7D?project=camunda-saas-int

**Comment by oleschoenburg** [comment]
I've tried to understand how this can happen but couldn't come up with any explanation.
Maybe we just adjust logging such that log the serialized form if deserialization fails.

**Comment by nicpuppa** [comment]
Happened again on a dev cluster in camunda cloud ([internal](https://console.cloud.google.com/errors/detail/CJPttavl2KffJw;service=zeebe;time=P7D?project=camunda-cloud-240911))

**Comment by koevskinikola** [comment]
Happened again on a prod cluster in Camunda Cloud.
* [GCP logs](https://console.cloud.google.com/logs/query;query=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22camunda-saas-prod%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-2%22%0Aresource.labels.location%3D%22northamerica-northeast2%22%0Aresource.labels.cluster_name%3D%22worker-5%22%0Aresource.labels.namespace_name%3D%22df66951e-6578-4c86-a985-8beb421c188d-zeebe%22;cursorTimestamp=2024-01-12T22:08:23.875462119Z;startTime=2024-01-12T21:09:01.773Z;endTime=2024-01-12T22:09:01.773Z?project=camunda-saas-prod) (internal).
* [JSON logs](https://drive.google.com/file/d/13aGCS4zGMQTRrTXlrJmZK48cE-hms7FF/view?usp=sharing) (internal).

**Comment by berkaycanbc** [comment]
Happened again and created a duplicate issue https://github.com/camunda/zeebe/issues/16128.

**Comment by oleschoenburg** [comment]
See https://github.com/camunda/zeebe/issues/16128 where we now have more detailed information. The topology that we couldn't deserialize was this:

```
clusterTopology {
  version: 1
  members {
    key: "0"
    value {
      lastUpdated {
        seconds: -31557014167219200
      }
      state: ACTIVE
      partitions {
        key: 1
        value {
          state: ACTIVE
          priority: 3
        }
      }
      partitions {
        key: 2
        value {
          state: ACTIVE
          priority: 1
        }
      }
      partitions {
        key: 3
        value {
          state: ACTIVE
          priority: 2
        }
      }
    }
  }
  members {
    key: "1"
    value {
      lastUpdated {
        seconds: -31557014167219200
      }
      state: ACTIVE
      partitions {
        key: 1
        value {
          state: ACTIVE
          priority: 2
        }
      }
      partitions {
        key: 2
        value {
          state: ACTIVE
          priority: 3
        }
      }
      partitions {
        key: 3
        value {
          state: ACTIVE
          priority: 1
        }
      }
    }
  }
  members {
    key: "2"
    value {
      lastUpdated {
        seconds: -31557014167219200
      }
      state: ACTIVE
      partitions {
        key: 1
        value {
          state: ACTIVE
          priority: 1
        }
      }
      partitions {
        key: 2
        value {
          state: ACTIVE
          priority: 2
        }
      }
      partitions {
        key: 3
        value {
          state: ACTIVE
          priority: 3
        }
      }
    }
  }
  lastChange {
  }
}
```

and the stacktrace is the same:
```
Caused by: java.lang.IllegalStateException: Unknown status: CHANGE_STATUS_UNKNOWN
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.toChangeStatus(ProtoBufSerializer.java:687) ~[zeebe-cluster-topology-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decodeCompletedChange(ProtoBufSerializer.java:352) ~[zeebe-cluster-topology-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decodeClusterTopology(ProtoBufSerializer.java:116) ~[zeebe-cluster-topology-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decode(ProtoBufSerializer.java:79) ~[zeebe-cluster-topology-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
```

**Comment by berkaycanbc** [comment]
Happened again:

* [GCP logs link (internal)](https://console.cloud.google.com/logs/query;query=error_groups.id%3D%22CM6G3rufvcTC8AE%22%0AlogName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.namespace_name%3D%222f7977a9-1ba7-44c7-8307-d7ef50e0ae57-zeebe%22%0Aresource.labels.location%3D%22us-central1%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-2%22%0Aresource.labels.cluster_name%3D%22prod-worker-3%22%0Aresource.labels.project_id%3D%22camunda-cloud-240911%22;cursorTimestamp=2024-01-30T07:17:57.938087773Z;startTime=2024-01-30T06:48:27.948Z;endTime=2024-01-30T07:48:27.948Z?project=camunda-cloud-240911)
* [GCP logs (internal)](https://drive.google.com/file/d/1uiDv04NCPL83J3A5nJ3c63AR8kKMzl2v/view)
* Zeebe version: `8.4.1`
* Cluster: Enterprise Trial Cluster

**Comment by michaelholtermann** [comment]
Hello,

I think we are affected as well when upgrading a Broker instance from 8.3.5 to 8.4.0:
```
Caused by: java.lang.IllegalStateException: Unknown status: CHANGE_STATUS_UNKNOWN
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.toChangeStatus(ProtoBufSerializer.java:677) ~[zeebe-cluster-topology-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decodeCompletedChange(ProtoBufSerializer.java:342) ~[zeebe-cluster-topology-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decodeClusterTopology(ProtoBufSerializer.java:109) ~[zeebe-cluster-topology-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.topology.serializer.ProtoBufSerializer.decode(ProtoBufSerializer.java:78) ~[zeebe-cluster-topology-8.4.0.jar:8.4.0]
```

Our settings:
```
            - name: ZEEBE_BROKER_CLUSTER_PARTITIONSCOUNT
              value: "3"
            - name: ZEEBE_BROKER_CLUSTER_CLUSTERSIZE
              value: "3"
            - name: ZEEBE_BROKER_CLUSTER_REPLICATIONFACTOR
              value: "3"
```

The result looks like this, screenshot from Zeebe Simple Monitor (we named the Zeebe Broker "process-control-tower-broker"):
![image](https://github.com/camunda/zeebe/assets/83571016/eb4637ef-318e-4558-a126-02ff3f3e3107)

Apparently, the followers on the same node cannot be started.

I can reproduce this issue, so I'd be happy to provide more details if required. For now, I'll stay with 8.3.5.

**Comment by npepinpe** [comment]
@oleschoenburg - could this lead to partitions becoming inactive? If so, we should re-assess the severity

**Comment by oleschoenburg** [comment]
No, the error only occurs in a request handler which means the request is not handled but the error is only logged, not escalated in any way: 
> Uncaught exception in '' in phase 'STARTED'. Continuing with next job.


@michaelholtermann that some brokers are inactive must have a different reason. I'd suggest to search the logs for different errors and report them in a new issue.

**Comment by rodrigo-lourenco-lopes** [comment]
Happened again in https://console.cloud.google.com/errors/detail/CM6G3rufvcTC8AE;service=zeebe;time=P7D?project=camunda-saas-int

**Comment by oleschoenburg** [comment]
@deepthidevaki and I found the root cause and it's a fun one :partying_face: 

In the two cases we've checked, this occurred during a rolling update from 8.3 to 8.4. While dynamic cluster configuration is not really supported in 8.3, at least some code made it into 8.3.0 (and later patches). On the broker side, everything was disabled with an experimental feature flag. However, the embedded gateway always starts a `GatewayClusterTopologyService` which runs the gossip service. If the entire cluster is 8.3, this isn't causing any issues.

But between 8.3 and 8.4 we made some breaking changes to the protobuf definitions. During a rolling update, new brokers in 8.4 recognize the rolling update, initialize the topology and gossip that around with the new protobuf definition. An old embedded gateway in 8.3 would be able to read this but then gossip around the same state serialized using an old protobuf definition. Particularly, the old protobuf includes the `ClusterChangePlan` as field 3 which is always set. The new broker uses field 3 for `CompletedChanges`. So as a result, the new broker would find that field 3 is set but none of the expected fields of `CompletedChanges` exist. Hence this bug: `lastChanges` is set but the status is `CHANGE_STATUS_UNKNOWN`.

Since we can safely assume that there will be no cluster configuration changes during a rolling update, this does not cause further issues apart from the noisy error log. It resolves itself once the rolling update is finished.

The only proper fix would be to go back to 8.3 and fix the embedded gateway to not start the gossip service. Since this is difficult to do and would only solve the noise, we'd rather accept the noise, acknowledge the error and wait for it to not show up again.
